server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java

```diff
@@ -129,42 +135,86 @@ private CurrentWsResponse toWsResponse(DbSession dbSession, String userLogin) {
-    return homepage.build();
+
+    return buildHomepage(dbSession, user.getHomepageType(), user.getHomepageParameter())
+      .orElse(defaultHomepage());
   }
 
-  private void setHomepageParameter(DbSession dbSession, String homepageType, @Nullable String homepageParameter, CurrentWsResponse.Homepage.Builder homepage) {
+  private Optional<CurrentWsResponse.Homepage> buildHomepage(DbSession dbSession, String homepageType, @Nullable String homepageParameter) {
+
     if (PROJECT.toString().equals(homepageType)) {
-      checkState(homepageParameter != null, HOMEPAGE_PARAMETER_SHOULD_NOT_BE_NULL);
-      ComponentDto component = dbClient.componentDao().selectByUuid(dbSession, homepageParameter)
-        .or(() -> {
-          throw new IllegalStateException(format("Unknown component '%s' for homepageParameter", homepageParameter));
-        });
-      homepage.setComponent(component.getKey());
-      setNullable(component.getBranch(), homepage::setBranch);
-      return;
+      return projectHomepage(dbSession, homepageType, homepageParameter);
     }
+
     if (APPLICATION.toString().equals(homepageType) || PORTFOLIO.toString().equals(homepageType)) {
-      checkState(homepageParameter != null, HOMEPAGE_PARAMETER_SHOULD_NOT_BE_NULL);
-      ComponentDto component = dbClient.componentDao().selectByUuid(dbSession, homepageParameter)
-        .or(() -> {
-          throw new IllegalStateException(format("Unknown component '%s' for homepageParameter", homepageParameter));
-        });
-      homepage.setComponent(component.getKey());
-      return;
+      return applicationAndPortfolioHomepage(dbSession, homepageType, homepageParameter);
     }
+
     if (ORGANIZATION.toString().equals(homepageType)) {
-      checkState(homepageParameter != null, HOMEPAGE_PARAMETER_SHOULD_NOT_BE_NULL);
-      OrganizationDto organization = dbClient.organizationDao().selectByUuid(dbSession, homepageParameter)
-        .orElseThrow(() -> new IllegalStateException(format("Unknown organization '%s' for homepageParameter", homepageParameter)));
-      homepage.setOrganization(organization.getKey());
+      return organizationHomepage(dbSession, homepageType, homepageParameter);
+    }
+
+    return of(CurrentWsResponse.Homepage.newBuilder()
+      .setType(CurrentWsResponse.HomepageType.valueOf(homepageType))
+      .build());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> projectHomepage(DbSession dbSession, String homepageType, @Nullable String homepageParameter) {
+    if (homepageParameter == null) {
+      LOGGER.warn("Ignoring Illegal Sate (Project key not specified for while homepage if of type Project), defaulting.");
+      return empty();
+    }
+
+    Optional<ComponentDto> projectOptional = ofNullable(dbClient.componentDao().selectByUuid(dbSession, homepageParameter).orNull());
+    if (!projectOptional.isPresent()) {
+      LOGGER.warn("Ignoring Illegal Sate (Project not found while specified), defaulting.");
```

Comment:
I think we should not log a warning and the message should be different, even if we would